qa/tasks/cephfs/test_client_limits.py

```diff
@@ -179,6 +180,13 @@ def test_cap_acquisition_throttle_readdir(self):
         cap_acquisition_throttle_hit_count = self.perf_dump()['mds_server']['cap_acquisition_throttle']
         self.assertGreaterEqual(cap_acquisition_throttle_hit_count, 1)
 
+        # make sure that the throttle was reported in the events
+        ops_dump = self.fs.mds_asok(['dump_historic_ops'])
+        # flat-map events from the operation list
+        op_events = lambda op: map(lambda ev: ev.get('event', ""), op.get('type_data', {}).get('events', []))
+        dump_events = chain.from_iterable(map(op_events, ops_dump.get('ops', [])))
+        self.assertIn("cap_acquisition_throttle", dump_events)
```

Comment:
My python skills are limited - I didn't know checking the presence of a string in an `itertools.chain` object can be done this way.

Reply:
this just flattens the array or array of strings and then uses the `in` to see if what we are looking for is there

Reply:
> this flattens the array or array of strings into a single stream and then uses the `in` to see if what we are looking for is there. `lambda ev: ev.get('event', "")` extracts the event name from the event structure, then `op_events = lamda op: ...` maps the event name extractor to the list of events given an `op`, providing an iterable of event names extracted from a single operation record; this is mapped to the stream of `ops` taken from the json output.

I got that. I played around with this a bit

```
>>> op_events = lambda op: map(lambda ev: ev.get('event', ""), op.get('type_data', {}).get('events', []))
>>> dump_events = chain.from_iterable(map(op_events, j.get('ops', [])))
>>> dump_events
<itertools.chain object at 0x7fc508955d00>
```

My questions was related to using (the assert check). I tried

```
>>> "replying" in dump_events
False
```

The json did have `replying` in the event list and `dump_events` is of type `<itertools.chain object` - therefore the question.

Reply:
OH I'm sorry I haven't understood your question.
Well, that's why I'm looking for the ways to get this test run :D

I've also used the interpreter to craft this piece of code given an actual output from the command, and it did work for me, finding the string with an `in`. Looking at the `assertIn` method I see it uses a simple `for in:` loop hence I hoped it would work. 

One of the reasons I think it could not work for you would be that you first printed the contents of `dump_events`. That would deplete the chain and subsequent enumerations would work as an empty sequence. To "reset" it you can repeat the line
```
>>> dump_events = chain.from_iterable(map(op_events, j.get('ops', [])))
```

Reply:
I think this discussion is a microcosm of why this sadly common type of Python code is absolutely horrible (please don't take personally Leonid). There's nothing intuitive or simple about it; it's mostly a mental exercise in trying to shrink code into as few lines as possible. Ultimately the code reader spends an inordinate amount of time trying to wrap their head around all the transformations on the maps/arrays/objects. I have actively discouraged the introduction of this type of code in the past.

Instead, what about:

```
ops_dump = self.fs.rank_tell(['dump_historic_ops'])
found = False
for op in ops_dump:
    for event in op['type_data']['events']:
        if event == 'cap_acquisition_throttle':
             fount = True
self.assertTrue(found)
```

This is trivial to understand and likely has better performance (not that matters for tests): there are no unnecessary intermediate objects created as with the successive transformations.

Reply:
I accept this POV, thanks for the feedback. I am a big fan of functional programming but will focus on readability.
In this case, I'd prefer to extract the suggested explicit nested loop into a function and call it in the assert statement, like

   self.assertTrue(self.eventInHistoricOps('cap_acquisition_throttle'))

re performance, in python3 the functional programming constructs are generators, which means that all the transformations get applied lazily, at least they should. Similar intermediate objects (iterators) most probably get created when iterating `for ... in`, so I'm not ready to fully agree with the performance argument without research, though it's true one can get into performance traps with functional programming.

Reply:
> re performance, in python3 the functional programming constructs are generators, which means that all the transformations get applied lazily, at least they should. Similar intermediate objects (iterators) most probably get created when iterating `for ... in`, so I'm not ready to fully agree with the performance argument without research, though it's true one can get into performance traps with functional programming.

Generators are emblematic of poorly designed programming constructs in Python. While appearing to have intuitive performance, they suffer from the poor choices surrounding Python's function/method call conventions which introduce **significant** overhead.  For example, see [benchf.txt](https://github.com/ceph/ceph/files/12217447/benchf.txt) which does a trivial benchmark of our two approaches:

```
functional: min/max/mean/stddev: 0.003109/0.003652/0.003274/0.000097
for loop: min/max/mean/stddev: 0.000886/0.001079/0.000945/0.000036
```

The `for` loop version is **71%** faster.

Python being a functional programming language is one of the biggest lies they've perpetuated. It is syntactic sugar for a vomit of bytecode that the CPU ultimately chokes on.

(As you may be able to tell now, I'm not an advocate for Python generally. However, with all this said, it's what we use for testing so I don't want to necessarily discourage using its useful features which help writing tests. I do think tests should be extremely easy and quick to read though, hence my protest.)

Reply:
Thanks for the test! I've corrected a few places which weren't equivalent, like the fact that lambdas were still using `reversed` arrays, and that the for loop code wasn't getting the `event` string property of events.
With the changes, the difference is smaller, but it's still true that a for loop is faster (unsurprisingly), in my runs x2:

```
» python3 benchf.py
functional: min/max/mean/stddev: 0.005257/0.007227/0.005531/0.000120
for loop: min/max/mean/stddev: 0.002411/0.002863/0.002521/0.000052
```

Your point is clear, hence my next push following your comments

<details><summary>changes to benchf</summary>

```
» diff -dC1 ~/Downloads/benchf.txt benchf.py 
*** /Users/leonidus/Downloads/benchf.txt        Mon Jul 31 21:59:11 2023
--- benchf.py   Mon Jul 31 22:09:05 2023
***************
*** 18,22 ****
              ops_dump = J
!             op_events = lambda op: map(lambda ev: ev.get('event', ""), reversed(op.get('type_data', {}).get('events', [])))
              #   flat-map events from the operation list
!             dump_events = chain.from_iterable(map(op_events, reversed(ops_dump.get('ops', []))))
              if "cap_acquisition_throttle" in dump_events:
--- 18,22 ----
              ops_dump = J
!             op_events = lambda op: map(lambda ev: ev['event'], op['type_data']['events'])
              #   flat-map events from the operation list
!             dump_events = chain.from_iterable(map(op_events, ops_dump['ops']))
              if "cap_acquisition_throttle" in dump_events:
***************
*** 36,38 ****
                  for event in op['type_data']['events']:
!                     if event == 'cap_acquisition_throttle':
                          found = True
--- 36,38 ----
                  for event in op['type_data']['events']:
!                     if event['event'] == 'cap_acquisition_throttle':
                          found = True
```
</details>

Reply:
It's an "Attach files..." button (but doesn't look like a button) just below the comment text box.